src/librustc/infer/canonical/mod.rs

```diff
@@ -74,6 +75,10 @@ pub struct CanonicalVarValues<'tcx> {
     pub var_values: IndexVec<CanonicalVar, Kind<'tcx>>,
 }
 
+/// Like CanonicalVarValues, but for use in places where a SmallVec is
+/// appropriate.
+pub type SmallCanonicalVarValues<'tcx> = SmallVec<[Kind<'tcx>; 8]>;
```

Comment:
Why not just use this everywhere...?

Reply:
Because `CanonicalVarValue` derives `Clone`, `Debug`, `PartialEq`, `Eq`, `Hash`, `RustcDecodable`, and `RustcEncodable`. In contrast, `SmallVec` doesn't define any of those. Also, I don't know what the impact of possible copying of `SmallVecs` (which are quite large, in terms of the number of bytes they take up on the stack) in lots of other places.